### ISSUE TITLE
Classical condition lifting breaks with specific `not` condition

### ISSUE THREAD

**Issue by swernli** [microsoft/qsharp-compiler#1112]
When experimenting with conditional statements and the conditional lifting, I found that certain equivalent statements will have one version work while the other does not. For example, setting `<ExecutionTarget>honeywell.test</ExecutionTarget>` to force classically controlled rewrite step to execute, the following code does not compile:
```qsharp
    @EntryPoint()
    operation SayHello() : Unit {
        use (q1, q2) = (Qubit(), Qubit());
        H(q1);
        H(q2);
        let r1 = M(q1);
        let r2 = M(q2);
        if (not (r1 == Zero and r2 == Zero)) {
            Message("Got here");
        }
    }
```
But this equivalent code DOES compile:
```qsharp
    @EntryPoint()
    operation SayHello() : Unit {
        use (q1, q2) = (Qubit(), Qubit());
        H(q1);
        H(q2);
        let r1 = M(q1);
        let r2 = M(q2);
        if (r1 == Zero and r2 == Zero){
            // NOOP
        } 
        else {
            Message("Got here");
        }
    }
```
The failure is actually in the C# compilation, because lifting ends up creating a C# function that is invalid:
```csharp
        public override Func<QVoid, QVoid> __Body__ => (__in__) =>
        {
            Microsoft__Quantum__ClassicalControl___0ef0a215c5fe4b52aaffda5a777b4168_ApplyIfElseR.Apply((r2, (Microsoft__Quantum__Canon___38ab4e8a4f274a06b0d58d979cd8ecd5_NoOp, QVoid.Instance), (_a0cff6d62c30402b91885de3aca53ef5_SayHello__, (r1, r2, q1, q2))));
#line hidden
            return QVoid.Instance;
        }
```
Note that this C# function uses the variables `r2`, `r1`, `q2`, and `q1` but never gets them from the input argument `__in__`, resulting in this compilation failure:
```
C:\Users\swern\Programming\test\LiftingBug\Program.qs(184,105): error CS0103: The name 'r2' does not exist in the current context [C:\Users\swern\Programming\test\LiftingBug\LiftingBug.csproj]
C:\Users\swern\Programming\test\LiftingBug\Program.qs(184,242): error CS0103: The name 'r1' does not exist in the current context [C:\Users\swern\Programming\test\LiftingBug\LiftingBug.csproj]
C:\Users\swern\Programming\test\LiftingBug\Program.qs(184,246): error CS0103: The name 'r2' does not exist in the current context [C:\Users\swern\Programming\test\LiftingBug\LiftingBug.csproj]
C:\Users\swern\Programming\test\LiftingBug\Program.qs(184,250): error CS0103: The name 'q1' does not exist in the current context [C:\Users\swern\Programming\test\LiftingBug\LiftingBug.csproj]
C:\Users\swern\Programming\test\LiftingBug\Program.qs(184,254): error CS0103: The name 'q2' does not exist in the current context [C:\Users\swern\Programming\test\LiftingBug\LiftingBug.csproj]
```
@ScottCarda-MS and @bettinaheim FYI

**Comment by ScottCarda-MS** [comment]
Is this at all related to #768?

**Comment by swernli** [comment]
I could be related, the but the symptoms are different. In #768, the generated C# will compile and run but the behavior at runtime will be incorrect. For this one, the compiler produces invalid C# code such that compilation fails.

**Comment by bettinaheim** [comment]
@ScottCarda-MS Could you please look into this bug?

**Comment by ScottCarda-MS** [comment]
Looks like the issue is that `KnownSymbols` is not getting properly propagated during the conditional restructuring phase of the 
Classical Control rewrite step. I'll keep looking into it.

**Comment by ScottCarda-MS** [comment]
Okay I believe I've found the issue. When we process the NOT conditions for restructuring, a new blank `else` block is made with not statements in it. This new block is also created without any known symbols. When comparing this to empty `else` blocks from source code, I see that the correct behavior is for the new `else` block to have the same known symbols as the other blocks in the conditional. I'll have a PR up soon with the fix.

**Comment by ScottCarda-MS** [comment]
This should be fixed now with #1117.
